### PR TITLE
Fix teardown issue of TCP interconnect

### PR DESCRIPTION
**Pull request by pengzhout**
Previously, for an interconnect connection, if no data are available at
sender peer, the sender sends a customized EOS packet to the receiver
and disables further send operations using shutdown(SHUT_WR), then
somehow, the sender closes the connection totally with close()
immediately and it counts on the kernel and TCP stack to guarantee the
data been transformed to the receiver. The problem is, on some platform,
if the connection is closed on one side, the TCP behave is undetermined,
the packets may be lost and receiver may report an unexpected error.

The correct way is sender blocks on the connection until receiver
getting the EOS packet and close its peer, then the sender can close
the connection safely.

### CODE REVIEW

**Comment by pivotal-ning-yu** [comment]
Looks like it's not easy to add test cases for this fix.

File: src/backend/cdb/motion/ic_tcp.c [hunk 1]
```diff
@@ -2502,15 +2516,29 @@ waitOnOutbound(ChunkTransportStateEntry *pEntry)
...
+				{
+					Assert(buf == 'S');
+					/* recv sent a stop message */
 					MPP_FD_CLR(conn->sockfd, &waitset);
 					/* we may have finished */
 					conn_count--;
 					continue;
 				}
-				if (count > 0 || (count < 0 && errno == EAGAIN))
+				else if (count < 0 && errno == EAGAIN)
```
[the remark this answers is not in the document]
Comment (pengzhout) [reply src/backend/cdb/motion/ic_tcp.c]: Not likely, the buf size is 1 in recv

File: src/backend/cdb/motion/ic_tcp.c [hunk 2]
```diff
@@ -2502,15 +2516,29 @@ waitOnOutbound(ChunkTransportStateEntry *pEntry)
...
+					if (enable)
+						elog(ERROR, "ssss");
+					MPP_FD_CLR(conn->sockfd, &waitset);
+					/* we may have finished */
+					conn_count--;
+					continue;
+				}
+				else if (count == 1)
+				{
+					Assert(buf == 'S');
```
[the remark this answers is not in the document]
Comment (pengzhout) [reply src/backend/cdb/motion/ic_tcp.c]: Hmm, make sense, fixed.

**Comment by pivotal-ning-yu** [comment]
Cool, LGTM

File: src/backend/cdb/motion/ic_tcp.c [hunk 3]
```diff
@@ -2502,15 +2516,20 @@ waitOnOutbound(ChunkTransportStateEntry *pEntry)
 
 				/* ready to read. */
 				count = recv(conn->sockfd, &buf, sizeof(buf), 0);
```
Comment (paul-guo-) [inline src/backend/cdb/motion/ic_tcp.c]: By the way, do we set any short timeout or keepalive option for those tcp sockets? If no, recv() could hang for a long time when the peer encounters kernel panic or NIC link status is down suddenly due to H/W issue, unless timeout or interrupt is encountered. I have been wondering that for blocking tcp recv() we should change to the more friendly logic below,

Loop with a total timeout {
    select/poll/epoll for read with a small timeout.
    call ML_CHECK_FOR_INTERRUPTS() or CHECK_FOR_INTERRUPTS() if needed.
    handle or break if needed.
}

Anyway this is probably not the thing that this patch should address.

File: src/backend/cdb/motion/ic_tcp.c [hunk 4]
```diff
@@ -2502,15 +2516,20 @@ waitOnOutbound(ChunkTransportStateEntry *pEntry)
...
 					MPP_FD_CLR(conn->sockfd, &waitset);
 					/* we may have finished */
 					conn_count--;
 					continue;
 				}
-				if (count > 0 || (count < 0 && errno == EAGAIN))
+				else if (count < 0 && errno == EAGAIN)
 					continue;
+				
 				/* Some other kind of error happened */
```
[the remark this answers is not in the document]
Comment (pengzhout) [reply src/backend/cdb/motion/ic_tcp.c]: existing code have done this.

File: src/backend/cdb/motion/ic_tcp.c (hunk 3 above)
Comment (pengzhout) [reply src/backend/cdb/motion/ic_tcp.c]: existing code have done this.

**Pull request merged**